### PR TITLE
PCR scanned followed by TAN Input does not trigger the PCR exchange screen

### PR DESCRIPTION
**Pull request by freshking**
## Description

Present overwrite notice when adding a new test via TAN and another test is already registered.

## Link to Jira

https://jira-ibs.wbs.net.sap/browse/EXPOSUREAPP-7143

### CODE REVIEW

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 1]
```diff
@@ -92,6 +92,32 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
+					didTapCloseButton: { [weak self] in
+						self?.navigationController?.dismiss(animated: true) {
+							feedbackHandler(false)
+						}
+					}
+				)
+
+				let footerViewController = FooterViewController(footerViewModel)
+				let topBottomViewController = TopBottomContainerViewController(topController: overwriteNoticeViewController, bottomController: footerViewController)
+				let nc = UINavigationController(rootViewController: topBottomViewController)
```
Comment (KaiTeuber) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: `TestOverwriteNoticeViewController` uses the protocol `dismissHandling` to handle that you should use `DismissHandlingNavigationController` instead.

**Review by KaiTeuber — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift (hunk 1 above)
Comment (freshking) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: @KaiTeuber I need the `UINavigationController` to show a navigation bar. `TopBottomContainerViewController` is presented and not pushed.

Comment (KaiTeuber) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: @freshking Just replace `UINavigationController` with `DismissHandlingNavigationController` that's all needed to respect the DismissHandling

**Review by KaiTeuber — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/TANInputViewController/TanInputViewModel.swift [hunk 2]
```diff
@@ -116,5 +114,22 @@ final class TanInputViewModel {
...
 	}
-
+	
+	private func registerPCRTest() {
+		isPrimaryButtonEnabled = false
+		isPrimaryBarButtonIsLoading = true
+		coronaTestService.registerPCRTest(teleTAN: text, isSubmissionConsentGiven: false) { [weak self] result in
+			switch result {
+			case let .failure(error):
+				// If teleTAN is incorrect, show Alert Controller
```
Comment (pascal-brause) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/TANInputViewController/TanInputViewModel.swift:124]: is in one of the following methods (like the alert) some error logged? If not, perhaps add some logs here...

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 3]
```diff
@@ -92,6 +92,32 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
+						self?.navigationController?.dismiss(animated: true) {
+							feedbackHandler(false)
+						}
+					}
+				)
+
+				let footerViewController = FooterViewController(footerViewModel)
+				let topBottomViewController = TopBottomContainerViewController(topController: overwriteNoticeViewController, bottomController: footerViewController)
+				let nc = DismissHandlingNavigationController(rootViewController: topBottomViewController)
+				self?.navigationController?.present(nc, animated: true)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:119]: Is there a reason why it is presented here and not pushed like it is after scanning a second test?

Comment (freshking) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:119]: It looked very weird to me pushing that view here.

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/TANInputViewController/TanInputViewModel.swift (hunk 2 above)
Comment (freshking) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/View/Controller/TANInputViewController/TanInputViewModel.swift:124]: Done

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift (hunk 3 above)
Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:119]: UX wanted it to be pushed and the whole submission flow to be dismissed when the close button is pushed.

I think it would be easiest and most consistent with the rest of the coordinator to call `showOverrideTestNotice` here and keep the logic that decides if it needs to be shown also in the coordinator. We would just need to extend the `CoronaTestQRCodeInformation` with the TAN case.

**Review by nickguendling — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift (hunk 3 above)
Comment (freshking) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift]: Done

**Review by naveeddotio — APPROVED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift [hunk 4]
```diff
@@ -92,6 +92,31 @@ class ExposureSubmissionCoordinator: NSObject, RequiresAppDependencies {
...
+					didTapCloseButton: { [weak self] in
+						self?.navigationController?.dismiss(animated: true) {
+							feedbackHandler(false)
+						}
+					}
+				)
+
+				let footerViewController = FooterViewController(footerViewModel)
+				let topBottomViewController = TopBottomContainerViewController(topController: overwriteNoticeViewController, bottomController: footerViewController)
+				self?.push(topBottomViewController)
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: I checked it out: it doesn't work, closes the flow when tapping on continue on the override notice screen. Thats why I suggested to extend the `CoronaTestQRCodeInformation` with the TAN case and use the same pattern as we do after scanning a QR code. It also shows the appropriate loading state on the confirmation screen as well as the error messages.

**Review by nickguendling — CHANGES_REQUESTED** [review]

File: src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift (hunk 4 above)
Comment (freshking) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: That is exactly the expected behavior!

Comment (nickguendling) [reply src/xcode/ENA/ENA/Source/Scenes/ExposureSubmission/ExposureSubmissionCoordinator.swift:118]: The same thing that happened before: the positive test result is shown so the user can submit the keys

**Review by nickguendling — CHANGES_REQUESTED** [review]
I like this approach, just have a couple of smaller suggestions :)

File: src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift [hunk 5]
```diff
@@ -441,7 +463,7 @@ class CoronaTestService {
 			.store(in: &subscriptions)
 	}
 
-	private func getRegistrationToken(
+	func getRegistrationToken(
```
Comment (nickguendling) [inline src/xcode/ENA/ENA/Source/Services/CoronaTestService/CoronaTestService.swift:466]: Is this change needed?

**Review by nickguendling — APPROVED** [review]

**Review by pascal-brause — APPROVED** [review]

File: src/xcode/ENA/ENAUITests/ENAUITests-Extensions.swift [hunk 6]
```diff
@@ -37,6 +37,7 @@ extension XCUIElement {
 		// Need XCTAssertEqual() here because we can append a message, this is not possible for XCTAssertTrue()
 		// swiftlint:disable:next xct_specific_matcher
 		XCTAssertEqual(self.waitForExistence(timeout: timeout), true, "Could not find XCUIElement: \(self) before tapping it.")
+		XCTAssertTrue(self.isEnabled)
```
Comment (freshking) [inline src/xcode/ENA/ENAUITests/ENAUITests-Extensions.swift:40]: @pascal-brause FYI
